File: solution/2192. All Ancestors of a Node in a Directed Acyclic Graph/2192_02.cpp
class Solution {
public:
    vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) {
        vector<vector<int>> adj(n);
        for (auto& e : edges)
            adj[e[0]].push_back(e[1]);
        
        vector<vector<int>> res(n);
        for (int i = 0; i < n; ++i) {
            dfs(i, i, adj, res);
        }
        return res;
    }
    
private:
    void dfs(int root, int node, vector<vector<int>>& adj, vector<vector<int>>& res) {
        for (int next : adj[node]) {
            if (!res[next].empty() && res[next].back() == root) continue;
            res[next].push_back(root);
            dfs(root, next, adj, res);
        }
    }
};

File: solution/2192. All Ancestors of a Node in a Directed Acyclic Graph/2192_02.py
class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        adj = defaultdict(list)
        res = [[] for _ in range(n)]
        for a, b in edges:
            adj[a].append(b)
        
        def dfs(node, root):
            for nxt in adj[node]:
                if res[nxt] and res[nxt][-1] == root: continue
                res[nxt].append(root)
                dfs(nxt, root)
        
        for i in range(n):
            dfs(i, i)
        return res
